File: leetcode-scala/src/dp/easy/IsSubsequence.scala
package dp.easy

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/8/28 13:51
 * @Description IsSubsequence
 */
object IsSubsequence {

    /**
     * 548 ms   5.00%
     * 50MB     71.43%
     * @param s
     * @param t
     * @return
     */
    def isSubsequence(s: String, t: String): Boolean = {
        if (s == null || s.length == 0) return true
        if (t == null || s.length > t.length) return false

        val c = s.charAt(0)
        val i = t.indexOf(c)
        i != -1 && isSubsequence(s.substring(1), t.substring(i + 1))
    }

    /**
     * 492ms    80.00%
     * 50.1MB   35.71%
     * @param s
     * @param t
     * @return
     */
    def isSubsequence2(s: String, t: String): Boolean = {
        var i = 0
        var j = 0
        while (i < s.length && j < t.length) {
            if (s(i) == t(j)) {
                i += 1
            }
            j += 1
        }
        i >= s.length
    }
}

File: leetcode-java/src/classify/dp/IsSubsequence.java
package classify.dp;

public class IsSubsequence {

    /**
     * 0ms 100.00%
     * 48.3MB 94.55%
     */
    public boolean isSubsequence(String s, String t) {
        int index = -1;

        for (char c : s.toCharArray()) {
            index = t.indexOf(c, index + 1);
            if (index == -1) {
                return false;
            }
        }

        return true;
    }

    /**
     * 1ms      81.62%
     * 355.8MB  99.81%
     * 双指针
     * @param s
     * @param t
     * @return
     */
    public boolean isSubsequence2(String s, String t) {
        int sIndex = 0, tIndex = 0;
        while (sIndex < s.length()) {
            while (tIndex < t.length() && t.charAt(tIndex) != s.charAt(sIndex)) {
                tIndex++;
            }
            if (tIndex >= t.length()) {
                return false;
            }
            sIndex++;
            tIndex++;
        }

        return true;
    }
}
